indexer/services/ender/src/lib/kafka-publisher.ts

```diff
@@ -43,34 +55,135 @@ export class KafkaPublisher {
   }
 
   public addEvent(event: ConsolidatedKafkaEvent) {
-    switch (event.topic) {
+    this.getMessages(event.topic)!.push(event.message);
+  }
+
+  /**
+   * Helper function to get messages for a given topic.
+   *
+   * @param kafkaTopic
+   * @private
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private getMessages(kafkaTopic: KafkaTopics): any[] | undefined {
+    switch (kafkaTopic) {
       case KafkaTopics.TO_WEBSOCKETS_SUBACCOUNTS:
-        this.subaccountMessages.push(event.message);
-        break;
+        return this.subaccountMessages;
       case KafkaTopics.TO_WEBSOCKETS_TRADES:
-        this.tradeMessages.push(event.message);
-        break;
+        return this.tradeMessages;
       case KafkaTopics.TO_WEBSOCKETS_MARKETS:
-        this.marketMessages.push(event.message);
-        break;
+        return this.marketMessages;
       case KafkaTopics.TO_WEBSOCKETS_CANDLES:
-        this.candleMessages.push(event.message);
-        break;
+        return this.candleMessages;
       case KafkaTopics.TO_VULCAN:
-        this.vulcanMessages.push(event.message);
-        break;
+        return this.vulcanMessages;
       default:
         throw new Error('Invalid Topic');
     }
   }
 
   /**
-   * Sort trade events by block height, transaction index, and event index in ascending order,
-   * where the first trade event should be the earliest trade in the block.
+   * Extract fills from subaccount message contents.
+   *
+   * @param contents
+   * @private
+   */
+  private extractFills(contents: string): FillSubaccountMessageContents[] | undefined {
+    try {
+      const subaccountContents:
+      SubaccountMessageContents = JSON.parse(contents) as SubaccountMessageContents;
+      return subaccountContents.fills;
+    } catch (error) {
+      throw new ParseMessageError(`Error parsing SubaccountMessageContents JSON: ${error}`);
+    }
+  }
+
+  private replaceFills(contents: string, fills: FillSubaccountMessageContents[]): string {
+    try {
+      const subaccountContents:
+      SubaccountMessageContents = JSON.parse(contents) as SubaccountMessageContents;
+      subaccountContents.fills = fills;
+      return JSON.stringify(subaccountContents);
+    } catch (error) {
+      throw new ParseMessageError(`Error parsing SubaccountMessageContents JSON: ${error}`);
+    }
+  }
+
+  /**
+   * Sort subaccountMessages that represent fills by block height, transaction index,
+   * and event index in ascending order per order id. Only keep the subaccount message if
+   * it represents the last fill event per order id, but make sure the subaccount message
+   * contents contains all individual fills from that block.
+   *
+   * Due to separate handlers for order fills, we can be sure that if a message is annotated
+   * with a fill, it should only contain data about a fill / order and not transfers.
```

Comment:
nit: Change this to note that the message should only contain a single fill / order.

Reply:
done